File: api/go/service.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"
)

type Response map[string]interface{}

var token = "GO_SECRET_TOKEN"
var notes = []map[string]string{}

func logging(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		log.Printf("%s %s %s %dms", r.Method, r.RequestURI, r.RemoteAddr, time.Since(start).Milliseconds())
	})
}

func checkAuth(r *http.Request) bool {
	auth := r.Header.Get("Authorization")
	if auth == "" || len(auth) < 8 || auth[:7] != "Bearer " {
		return false
	}
	return auth[7:] == token
}

func handleAuthLogin(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	var body struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(Response{"error": "invalid json"})
		return
	}
	if body.Username == "admin" && body.Password == "admin" {
		json.NewEncoder(w).Encode(Response{"token": token})
		return
	}
	w.WriteHeader(http.StatusUnauthorized)
	json.NewEncoder(w).Encode(Response{"error": "invalid credentials"})
}

func handleStatus(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(Response{"status": "go up", "timestamp": time.Now().UTC().Format(time.RFC3339)})
}

func handleEcho(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	if !checkAuth(r) {
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(Response{"error": "unauthorized"})
		return
	}
	var body struct {
		Message string `json:"message"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(Response{"error": "invalid json"})
		return
	}
	notes = append(notes, map[string]string{"message": body.Message, "created_at": time.Now().UTC().Format(time.RFC3339)})
	json.NewEncoder(w).Encode(Response{"engine": "go", "message": body.Message, "echoed": time.Now().UTC().Format(time.RFC3339)})
}

func handleNotes(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	if !checkAuth(r) {
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(Response{"error": "unauthorized"})
		return
	}
	json.NewEncoder(w).Encode(Response{"notes": notes})
}

func handleQuote(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(Response{"engine": "go", "quote": "Go is expressive, concise, clean, and efficient."})
}

func handleMetrics(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	fmt.Fprintf(w, "go_uptime_seconds %d\n", int(time.Now().Unix()))
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/status", handleStatus)
	mux.HandleFunc("/auth/login", handleAuthLogin)
	mux.HandleFunc("/echo", handleEcho)
	mux.HandleFunc("/notes", handleNotes)
	mux.HandleFunc("/quote", handleQuote)
	mux.HandleFunc("/metrics", handleMetrics)

	log.Println("Go API running at http://localhost:8081")
	if err := http.ListenAndServe(":8081", logging(mux)); err != nil {
		log.Fatal(err)
	}
}

File: api/python/service.py
from flask import Flask, request, jsonify
from flask_cors import CORS
from datetime import datetime
import logging

logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')
app = Flask(__name__)
CORS(app, resources={r"/api/*": {"origins": ["http://localhost", "http://localhost:8080"]}})

quotes = [
    'Python makes it easy to write prototypes quickly.',
    'Readability counts.',
    'Pythonic is better than non-pythonic.'
]

TOKEN = 'PYTHON_SECRET_TOKEN'

notes = []

@app.route('/status')
def status():
    logging.info('status hit')
    return jsonify({'status': 'python up', 'time': datetime.utcnow().isoformat() + 'Z', 'env': 'production'})

@app.route('/auth/login', methods=['POST'])
def auth_login():
    payload = request.get_json(force=True, silent=True) or {}
    if payload.get('username') == 'admin' and payload.get('password') == 'admin':
        return jsonify({'token': TOKEN})
    return jsonify({'error': 'invalid credentials'}), 401


def require_auth():
    auth = request.headers.get('Authorization', '')
    if not auth.startswith('Bearer '):
        return False
    return auth.split(' ')[1] == TOKEN

@app.route('/echo', methods=['POST'])
def echo():
    if not require_auth():
        return jsonify({'error': 'unauthorized'}), 401
    payload = request.get_json(force=True, silent=True) or {}
    message = payload.get('message')
    if not message or not isinstance(message, str):
        logging.warning('echo bad request: %s', payload)
        return jsonify({'error': 'message is required and must be a string'}), 400

    note = {'message': message, 'ts': datetime.utcnow().isoformat() + 'Z'}
    notes.append(note)
    response = {'engine': 'python', 'message': message, 'received': datetime.utcnow().isoformat() + 'Z'}
    logging.info('echo response: %s', response)
    return jsonify(response)

@app.route('/notes')
def list_notes():
    if not require_auth():
        return jsonify({'error': 'unauthorized'}), 401
    return jsonify({'notes': notes})

@app.route('/quote')
def quote():
    q = quotes[datetime.utcnow().second % len(quotes)]
    logging.info('quote hit: %s', q)
    return jsonify({'engine': 'python', 'quote': q})

@app.route('/metrics')
def metrics():
    return 'python_uptime_seconds %s\n' % (datetime.utcnow() - datetime(1970,1,1)).total_seconds(), 200, {'Content-Type': 'text/plain'}

@app.errorhandler(Exception)
def handle_error(e):
    logging.exception('Unhandled exception')
    return jsonify({'error': 'internal server error'}), 500

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=False)
